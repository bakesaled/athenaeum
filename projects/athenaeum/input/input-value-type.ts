export type AthInputValueType = 'number' | 'text' | null;
