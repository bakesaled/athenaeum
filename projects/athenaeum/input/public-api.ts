export * from './input.directive';
export * from './input.module';
export * from './input-value-type';
