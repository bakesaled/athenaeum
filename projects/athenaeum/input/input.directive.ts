import {
  Directive,
  ElementRef,
  forwardRef,
  HostListener,
  Input,
} from '@angular/core';
import { MAT_INPUT_VALUE_ACCESSOR } from '@angular/material/input';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { formatNumber } from '@angular/common';
import { AthInputValueType } from './input-value-type';

/**
 * A directive that adds helpful behavior to input and textarea elements.
 */
@Directive({
  selector: 'input[athInput], textarea[athInput]',
  exportAs: 'athInput',
  providers: [
    { provide: MAT_INPUT_VALUE_ACCESSOR, useExisting: AthInputDirective },
    {
      provide: NG_VALUE_ACCESSOR,
      useExisting: forwardRef(() => AthInputDirective),
      multi: true,
    },
  ],
  host: {
    class: 'ath-input-element',
    '[style.text-align]': '"right"',
  },
})
export class AthInputDirective {
  private val: string;

  /**
   * Value assigned to the element (MatFormField).
   */
  @Input()
  get value(): string {
    return this.val;
  }

  set value(value: string | null) {
    this.val = value;
    this.formatValue(value);
  }

  locale = 'en-US';

  /**
   * Decimal representation options, specified by a string in the following format:
   * `{minIntegerDigits}.{minFractionDigits}-{maxFractionDigits}`.
   */
  @Input()
  digitInfo: string;

  /**
   * Type of value: 'number' or 'text' (default: 'text');
   */
  @Input()
  valueType: AthInputValueType;

  constructor(private elementRef: ElementRef<HTMLInputElement>) {}

  @HostListener('blur')
  onBlur(): void {
    if (this.val) {
      this.formatValue(this.val);
    }
  }

  @HostListener('focus')
  onFocus(): void {
    this.unformatValue();
  }

  @HostListener('input', ['$event.target.value'])
  onInput(value): void {
    if (this.valueType === 'number') {
      this.val = this.unformatNumber(value);
    } else {
      this.val = value;
    }
    this.onChange(this.val);
  }

  writeValue(value: any): void {
    this.val = value;
    if (value) {
      this.formatValue(this.val);
    }
  }

  onChange(_): void {}

  registerOnChange(fn: (value: any) => void): void {
    this.onChange = fn;
  }

  registerOnTouched(): void {}

  unformatNumber(target: string): string {
    if (target) {
      return target.replace(/[^\d.-]/g, '');
    } else {
      return '';
    }
  }

  private formatValue(value: any): void {
    if (this.valueType === 'number') {
      this.elementRef.nativeElement.value = formatNumber(
        value,
        this.locale,
        this.digitInfo
      );
    }
  }

  private unformatValue(): void {
    if (this.valueType === 'number') {
      this.elementRef.nativeElement.value = this.unformatNumber(
        this.elementRef.nativeElement.value
      );
    }
  }
}
