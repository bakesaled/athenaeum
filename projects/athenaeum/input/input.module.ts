import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AthInputDirective } from './input.directive';
import { MatInputModule } from '@angular/material/input';

@NgModule({
  declarations: [AthInputDirective],
  imports: [CommonModule, MatInputModule],
  exports: [AthInputDirective],
})
export class AthInputModule {}
