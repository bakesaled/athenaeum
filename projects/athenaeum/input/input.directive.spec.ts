import { AthInputDirective } from './input.directive';
import { Component } from '@angular/core';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { MatInputModule } from '@angular/material/input';
import { FormsModule } from '@angular/forms';
import { MatFormFieldModule } from '@angular/material/form-field';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { By } from '@angular/platform-browser';

@Component({
  template: `
    <mat-form-field>
      <input matInput athInput placeholder="Enter text here" />
    </mat-form-field>
  `,
})
class AthInputTestComponent {}

describe('InputDirective', () => {
  let fixture: ComponentFixture<AthInputTestComponent>;
  let componentInstance: AthInputTestComponent;
  let containerElement: HTMLElement;
  let directive: AthInputDirective;

  beforeEach(
    waitForAsync(() => {
      TestBed.configureTestingModule({
        declarations: [AthInputDirective, AthInputTestComponent],
        imports: [
          MatInputModule,
          FormsModule,
          MatFormFieldModule,
          NoopAnimationsModule,
        ],
      });

      fixture = TestBed.createComponent(AthInputTestComponent);
      componentInstance = fixture.debugElement.componentInstance;
      containerElement = fixture.debugElement.query(By.css('input'))
        .nativeElement;
      const debugEl = fixture.debugElement.query(
        By.directive(AthInputDirective)
      );
      directive = debugEl.injector.get(AthInputDirective);
    })
  );

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should create an instance', () => {
    expect(directive).toBeTruthy();
  });

  it('should create an instance', () => {
    expect(fixture).toBeTruthy();
  });

  it('should render', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should unformat number', () => {
    let result = directive.unformatNumber('1,234');
    expect(result).toBe('1234');
    result = directive.unformatNumber('1,234.56');
    expect(result).toBe('1234.56');
    result = directive.unformatNumber('123');
    expect(result).toBe('123');
  });

  it('should return 0 if unformatting null', () => {
    let result = directive.unformatNumber(null);
    expect(result).toBe('');
    result = directive.unformatNumber(undefined);
    expect(result).toBe('');
  });
});
