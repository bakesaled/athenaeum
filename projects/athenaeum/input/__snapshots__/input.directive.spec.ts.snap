// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputDirective should render 1`] = `
<ng-component>
  <mat-form-field
    class="mat-form-field mat-primary"
  >
    <div
      class="mat-form-field-wrapper"
    >
      <div
        class="mat-form-field-flex"
      >
        
        
        <div
          class="mat-form-field-infix"
        >
          <input
            athinput=""
            class="ath-input-element mat-input-element mat-form-field-autofill-control"
            matinput=""
            placeholder="Enter text here"
          />
          <span
            class="mat-form-field-label-wrapper"
          >
            
          </span>
        </div>
        
      </div>
      
      <div
        class="mat-form-field-subscript-wrapper"
      >
        
        
      </div>
    </div>
  </mat-form-field>
</ng-component>
`;
